WooCommerce/Classes/POS/ViewModels/ItemListViewModel.swift

```diff
@@ -42,25 +27,135 @@ final class ItemListViewModel: ObservableObject {
-            items.removeAll()
-            items = newItems
-            state = .loaded
+            if newItems.count == 0 {
+                let itemListEmpty = EmptyModel(title: Constants.emptyProductsTitle,
+                                                  subtitle: Constants.emptyProductsSubtitle,
+                                                  hint: Constants.emptyProductsHint,
+                                                  buttonText: Constants.emptyProductsButtonTitle)
+                state = .empty(itemListEmpty)
+            } else {
+                // Only clears in-memory items if the `do` block continues, otherwise we keep them in memory.
+                items.removeAll()
+                items = newItems
+                state = .loaded(items)
+            }
         } catch {
             DDLogError("Error on reload while updating product data: \(error)")
-            state = .error
+            let itemListError = ErrorModel(title: Constants.failedToLoadTitle,
+                                      subtitle: Constants.failedToLoadSubtitle,
+                                      buttonText: Constants.failedToLoadButtonTitle)
+            state = .error(itemListError)
+        }
+    }
+}
+
+extension ItemListViewModel {
+    enum ItemListState: Equatable {
+        case empty(EmptyModel)
+        // TODO:
+        // Differentiate between loading on entering POS mode and reloading, as the
+        // screens will be different:
+        // https://github.com/woocommerce/woocommerce-ios/issues/13286
+        case loading
+        case loaded([POSItem])
+        case error(ErrorModel)
+
+        // Equatable conformance for testing:
+        static func == (lhs: ItemListViewModel.ItemListState, rhs: ItemListViewModel.ItemListState) -> Bool {
+            switch (lhs, rhs) {
+            case (.empty(let lhsItems), .empty(let rhsItems)):
+                return lhsItems == rhsItems
+            case (.loading, .loading):
+                return true
+            case (.loaded(let lhsItems), .loaded(let rhsItems)):
+                return lhsItems.map { $0.itemID } == rhsItems.map { $0.itemID }
+            case (.error(let lhsError), .error(let rhsError)):
+                return lhsError == rhsError
+            default:
+                return false
+            }
         }
     }
+
+    struct ErrorModel: Equatable {
+        let title: String
+        let subtitle: String
+        let buttonText: String
+    }
+
+    struct EmptyModel: Equatable {
+        // TODO:
+        // Differentiate between empty with no products vs empty with no eligible products
+        // https://github.com/woocommerce/woocommerce-ios/issues/12815
+        // https://github.com/woocommerce/woocommerce-ios/issues/12816
+        let title: String
+        let subtitle: String
+        let hint: String
+        let buttonText: String
+    }
```

Comment:
Those seem extremely similar, I'm thinking we could use a single model for both since there's no need to pass something like `Error` to the UI and we may use only strings and images. Thoughts?

Reply:
> If business logic and UI-wise we differentiate between error (cannot fetch), and empty (no eligible product) states, I think it's fine not to over-optimize and allow them to be described by different structs.

That's a good thought, while what we present in the UI might be similar, these cover different cases and will have different business logic in the CTAs, so for now let's keep them separate.